adding custom logging test + local runtimes-common integration test

balopat opened the pull request:
* This the first half of #5 enabling [custom logging tests](https://github.com/GoogleCloudPlatform/runtimes-common/tree/master/integration_tests/#custom-logging). The standard logging tests are going to be in the next PR. 
   * Important to note: https://github.com/GoogleCloudPlatform/runtimes-common/pull/308 is a prerequisite for this, otherwise the driver is not able to pick up the project id from the docker container (as it's using gcloud for that which is currently not setup in the docker container with the right information). However, for now I published my branch version to the java-runtimes-test docker registry that can serve as our version of the driver until we negotiate the PR with @nkubala.
  * closes #126 
  * closes #145 
  * closes #140

TODO left: 

- [x] fix local runtimes-common integration tests with local cloud builder (if possible)
- [x] fix logging test timeout issue (recently came up) 
- [x] final review and refactor (opened new ticket for a larger redesign: #148)

aslo commented on scripts/local_integration_test.sh:
```diff
@@ -44,48 +90,66 @@ pushd $deployDir
+docker build -t $APP_IMAGE . || gcloud docker -- build -t $APP_IMAGE . || usage "Error building test-app image from base image \"${imageUnderTest}\", please make sure it exists!"
+
+
+if [[ "$WITH_GCP" ]]; then
```
I don't think it makes sense to me to add these integration tests to `the local_integration_test.sh` script. I think we originally added this script in order to test things that were outside the scope of the runtimes_common integraion testing framework (namely, container shutdown hooks).

I think if we wanted to make it easier to run the runtimes_common tests, a better way to do it would be to just simulate a cloudbuild execution locally, using the `container-builder-local` emulator that was just released as part of the Cloud SDK (see [release notes](https://cloud.google.com/sdk/docs/release-notes#google_cloud_container_builder)).

balopat replied:
oooo, I'll test out this ``container-build-local``, thanks for the suggestion! 
I created a separate script first btw, but then I figured that they were so similar that I integrated it back into the local_integration_test.sh. It's really handy to have as part of the workflow. But if container-build-local can do the magic, I'll have a look at that as that could remove the duplication around the arguments of the runtimes-common driver for example (one place here, the other place is the yaml file).

balopat replied:
I will do this in #126 though, and then separate it out from this pull request.

aslo commented on test-application/src/main/java/com/google/cloud/runtimes/config/MockGCPConfiguration.java:
```diff
@@ -0,0 +1,53 @@
+import static org.mockito.Mockito.mock;
+
+@Configuration
+@Profile("mock-gcp")
```
I'm skeptical about having these separate profiles for gcp and mock-gcp. Is it possible to write a single test app implementation, and have the environment in which the test app runs handle the responsibility of passing in google credentials?

balopat replied:
One of the reasons I introduced this was that wihtout mock-gcp - the ``local_integration_test.sh`` (and I mean the shutdown log testing too, not just the local runtime-commons test) would need to pass in the GCP creds, otherwise the app doesn't start up. Is it okay to expect a valid GCP setup for the user in the local_integration_test.sh?

aslo replied:
Maybe the `local_integration_test.sh` script could check for the presence of the `GOOGLE_CREDENTIALS` environment variable, and either fail or warn the user if it looks like the environment wasn't properly set up. What do you think about that?

balopat replied:
It's doable, the only thing I have against that is that this decision will force someone to setup a service account even when they start their development...which might not be a too strong issue but we'll have to document it then.

balopat replied:
@meltsufin what do you think?

meltsufin replied:
Can we just not invoke this logging test when running locally? Another option would be to have a separate test-app that doesn't depend on GCP for local testing. Otherwise, I don't really mind the mock-gcp profile.

balopat replied:
The problem is that I am using Spring to inject the services, so when the beans get created, they need the credentials and thus _the application fails to start_ without the creds - that's why I went down the road to create the two profiles. I just had an idea that I'll check out: I should try to make the components ``@Lazy``.
Re: skipping the GCP related tests locally: if we combine @aslo's idea of separating the runtimes-common tests out into a separate script, then this will naturally skip calling them in the local integration tests.
So, if I can make ``@Lazy`` work, we have a "single test app implementation" (without need for profiles), the app starts up and we'll have the runtimes-common tests separated, which might need the GCP creds.

meltsufin commented on test-application/src/main/java/com/google/cloud/runtimes/LoggingTestController.java:
```diff
@@ -0,0 +1,84 @@
+import com.google.cloud.logging.Logging;
+import com.google.cloud.logging.Payload;
+import com.google.cloud.logging.Severity;
+import com.google.cloud.runtimes.stackdriver.StackDriverMonitoringService;
```
Unused import

meltsufin commented on test-application/src/main/java/com/google/cloud/runtimes/LoggingTestController.java:
```diff
@@ -0,0 +1,84 @@
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+import sun.util.logging.resources.logging;
```
Unusued import

meltsufin commented on test-application/src/main/java/com/google/cloud/runtimes/LoggingTestController.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    @RequestMapping(path = "/logging_custom", method = POST)
+    public String handleMonitoringRequest(@RequestBody LoggingTestRequest loggingTestRequest) throws IOException, InterruptedException {
```
`handleLoggingRequest`?

meltsufin commented on test-application/src/main/java/com/google/cloud/runtimes/LoggingTestController.java:
```diff
@@ -0,0 +1,84 @@
+                .setResource(MonitoredResource.newBuilder("global").build())
+                .build();
+        entries.add(entry);
+        logging.write(entries);
```
Instead of using the Logging client lib directly, we should be able to use the JUL appender. Can you also look at #72 that has been put on the back-burner, and reconcile it with what you're doing in this PR? Ideally, we would like to close out #72 as well, but I think there are some unresolved complications there with setting up JUL.

cassand replied:
I think that the Custom Logging Test should use the client library and make an explicit call to Stackdriver Logging (by creating only one LogEntry), whereas the Standard Logging Test should redirect the output of JUL to Stackdriver.

balopat replied:
I agree with @cassand - I was planning to use JUL for standard logging tests and Logging library directly for custom, just so that we have both covered. 
However, we are free to decide what we want to do, as we can setup custom logging with JUL too.

meltsufin replied:
I'm OK with Damien's suggestion.

meltsufin commented on test-application/src/main/java/com/google/cloud/runtimes/config/GCPConfiguration.java:
```diff
@@ -0,0 +1,35 @@
+
+@Configuration
+@Profile("gcp")
+public class GCPConfiguration {
```
`GcpConfiguration`

meltsufin commented on test-application/src/main/java/com/google/cloud/runtimes/config/MockGCPConfiguration.java:
```diff
@@ -0,0 +1,53 @@
+
+@Configuration
+@Profile("mock-gcp")
+public class MockGCPConfiguration {
```
`MockGcpConfiguration`

balopat commented:
@aslo @meltsufin can you please review again this PR? I addressed all of your concerns: 
- @aslo: removed the separate MockGCP profile and introduced container-builder-local for local testing
- @meltsufin: cleanup of unused imports and method rename (handleLoggingRequest) is done 

Thanks!

meltsufin commented on scripts/local_runtimes_common_integration_test.sh:
```diff
@@ -0,0 +1,111 @@
+  exit 1
+fi
+
+if [[ -z ${GOOGLE_APPLICATION_CREDENTIALS} ]]; then
```
The readme instructions don't mention `GOOGLE_APPLICATION_CREDENTIALS`.
Can we just leverage gcloud for the default credentials?

meltsufin commented on scripts/local_runtimes_common_integration_test.sh:
```diff
@@ -0,0 +1,111 @@
+
+docker rm -f $CONTAINER || echo "Integration-test-app container is not running, ready to start a new instance."
+
+# run app container locally to test shutdown logging
```
You're not really doing a shutdown logging test here.

balopat commented on test-application/pom.xml:
```diff
@@ -35,13 +35,33 @@
+    </dependency>
+    <dependency>
+      <groupId>com.google.api</groupId>
+      <artifactId>gax-grpc</artifactId>
```
good catch - let me double check & I will clarify in comments.
